### PR TITLE
[Model] Pooling model activation supports per request control by PoolingParams

### PR DESCRIPTION
**Pull request by noooop**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

Pooling model “activation” supports per request control by PoolingParams

- Consider only the dimensions and normalize parameters for embed models, 
- Consider only the activation parameter for classification (score) models
- Consider only the softmax parameter for PoolingType.ALL reward models
- Consider only the softmax/step_tag_id/returned_token_ids parameter for PoolingType.STEP reward models

Config priority, from low to high
1. vllm default in vllm/pooling_params.py _set_default_parameters
2. get_pooling_config (from sentencetransformer config) in vllm/config.py _init_pooler_config
3. verify_and_update_config in vllm/model_executor/models/config.py
4. override_pooler_config in vllm/config.py _init_pooler_config
5. PoolingParams.__init__ in vllm/pooling_params.py 

To make the control logic simpler and more intuitive, we need to refactor the entire pooler control logic as follows:

1. override_pooler_config is used to modify the default parameters of pooler_config   (_init_pooler_config & verify_and_update_config in vllm/config.py)
2. Allow users to set Pooler parameters via PoolingParams per request. All default parameters are merged from pooler_config into PoolingParams (PoolingParams.merge_default_parameters + PoolingParams.verify in vllm/pooling_params.py)
3. PoolingParams is used to per request control Pooler ( PoolingMethod& PoolerActivation) ( in vllm/model_executor/layers/pooler.py)

More specific Pooler
- EmbeddingPoolerHead + SimplePooler for embed models, 
- ClassifierPooler  for classification (score) models
- RewardPoolerHead + SimplePooler for PoolingType.ALL reward models
- RewardPoolerHead + StepPooler for PoolingType.STEP reward models

## Test Plan

- override_pooler_config -> pooler_config -> PoolingParams

pytest -s -vvv  tests/test_pooling_params.py

- normalize for embed api
```
pytest -s -vvv tests/models/language/pooling/test_override_pooler_config.py::test_embed_models_using_normalize
pytest -s -vvv tests/entrypoints/llm/test_embedding.py::test_pooling_params
pytest -s -vvv tests/entrypoints/openai/test_embedding.py::test_normalize
```

- activation for classify api
```
pytest -s -vvv tests/models/language/pooling/test_override_pooler_config.py::test_classify_models_using_activation
pytest -s -vvv tests/entrypoints/llm/test_classification.py::test_pooling_params
pytest -s -vvv tests/entrypoints/openai/test_classification.py::test_activation

```

- activation for score api
```
pytest -s -vvv tests/entrypoints/llm/test_score.py::test_pooling_params
pytest -s -vvv tests/entrypoints/openai/test_score.py::test_activation
pytest -s -vvv tests/entrypoints/openai/test_rerank.py::test_activation
```

- softmax for reward api
```
pytest -s -vvv tests/models/language/pooling/test_override_pooler_config.py::test_reward_models_using_softmax
pytest -s -vvv tests/entrypoints/llm/test_reward.py::test_pooling_params
```

## Test Result

pass

## Known Issues

We may need to refactor vllm/entrypoints/openai/serving_pooling.py because it uses EmbeddingCompletionRequest, which may deviate from the requirements of reward models.

```
PoolingCompletionRequest = EmbeddingCompletionRequest
PoolingChatRequest = EmbeddingChatRequest
PoolingRequest = Union[PoolingCompletionRequest, PoolingChatRequest]
```

### CODE REVIEW

**Comment by noooop** [comment]
@DarkLight1337 

after #20527 

Bert like ForSequenceClassification, for example, papluca/xlm-roberta-base-language-detection is still not controlled by override_pooler_config.

On the other hand, the model converted using as_seq_cls_model, for example,  jason9693/Qwen2.5-1.5B-apeach is controlled by override_pooler_config. 

Do we need to unify this?

**Comment by DarkLight1337** [comment]
Yes, `override_pooler_config` should be applied in all cases

**Comment by maxdebayser** [comment]
I wonder how that relates to the context extension PR. Would it perhaps make a difference to disable the normalize and softmax for the chunk processing and then apply them on the normalized results?

Another question is whether this would prevent us from using cuda graphs on the pooler.

**Comment by noooop** [comment]
> I wonder how that relates to the context extension PR. Would it perhaps make a difference to disable the normalize and softmax for the chunk processing and then apply them on the normalized results?
> 
> Another question is whether this would prevent us from using cuda graphs on the pooler.

This PR is unrelated to the context extension PR.

The pooler uses too many branch computations, making it inherently unsuitable for CUDA graphs. Using CUDA graphs on the backbone network is already efficient enough.

**Comment by noooop** [comment]
This PR is still in the WIP stage, and I'm not sure what the final goal is. I'll call you to review once I'm done.

**Comment by noooop** [comment]
@DarkLight1337 @maxdebayser 

Ready for review

Since task can be set per request, there is no reason why normalize & activation cannot.

File: tests/entrypoints/llm/test_score.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import torch.nn.functional as F
+
+from vllm import LLM, PoolingParams
+from vllm.distributed import cleanup_dist_env_and_memory
+
+MODEL_NAME = "BAAI/bge-reranker-v2-m3"
+
+
+@pytest.fixture(autouse=True)
+def v1(run_with_both_engines):
```
Comment (maxdebayser) [inline tests/entrypoints/llm/test_score.py:18]: AFAIK `XLMRobertaForSequenceClassification` is not supported in V1 yet.

File: tests/entrypoints/llm/test_embedding.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from vllm import LLM, PoolingParams
+from vllm.distributed import cleanup_dist_env_and_memory
+
+MODEL_NAME = "intfloat/multilingual-e5-small"
+
+prompts = ["The chef prepared a delicious meal."]
+
+
+@pytest.fixture(autouse=True)
+def v1(run_with_both_engines):
```
Comment (maxdebayser) [inline tests/entrypoints/llm/test_embedding.py]: `BertModel` is not working in V1 yet.

**Review by maxdebayser — APPROVED** [review]
Overall it looks good to me.

**Comment by maxdebayser** [comment]
> This PR is unrelated to the context extension PR.

I know, I'm just curious.

> The pooler uses too many branch computations, making it inherently unsuitable for CUDA graphs. Using CUDA graphs on the backbone network is already efficient enough.

Well yes, now that you can choose between all the different tasks at runtime.

File: vllm/model_executor/layers/pooler.py [hunk 3]
```diff
@@ -670,12 +689,14 @@ def __init__(
...
         classifier: ClassifierFn,
         act_fn: PoolerActivation,
+        activation: bool = True,
     ) -> None:
         super().__init__()
 
         self.pooling = pooling
         self.classifier = classifier
         self.act_fn = act_fn
+        self.activation = activation
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]: We can simplify the code by using identity activation if `activation=False`. That way we can reduce the amount of branching

File: vllm/model_executor/layers/pooler.py [hunk 4]
```diff
@@ -489,43 +490,30 @@ def from_config(cls, pooler_config: ResolvedPoolingConfig) -> "PoolerHead":
...
+        return cls(pooler_config)
 
-    def __init__(self, activation: PoolerActivation) -> None:
+    def __init__(self, pooler_config: ResolvedPoolingConfig) -> None:
         super().__init__()
 
-        self.activation = activation
+        self.pooler_config = pooler_config
+        self.normalize = PoolerNormalize()
+        self.softmax = PoolerClassify()
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]: Factor this out into a method `_get_activation` so we can reuse the previous code

**Comment by noooop** [comment]
@DarkLight1337

Known Issues

pooler_config.softmax did not control classification (score) models previously , and also does not control classification (score) models now .

pooler_config.softmax is used only in the reward model,  e.g. Qwen2ForProcessRewardModel

embed models -> PoolerHead + PoolerNormalize
reward model -> PoolerHead + PoolerClassify
classification (score) models -> ClassifierPooler

How about splitting PoolerHead into NormalizePoolerHead & RewardPoolerHead?

embed models -> NormalizePoolerHead 
reward model -> RewardPoolerHead
classification (score) models -> ClassifierPooler

**Comment by DarkLight1337** [comment]
Why can't the current PoolerHead work in all these cases? Just pass in the correct activation type to it

**Comment by DarkLight1337** [comment]
We should avoid creating a complicated inheritance hierarchy.

**Comment by DarkLight1337** [comment]
IIRC there are currently some models that hardcode the activation type.

**Comment by DarkLight1337** [comment]
We should add some abstraction to the selection of activation function so user arguments are not silently ignored

**Comment by noooop** [comment]
embed models -> NormalizePoolerHead -> matryoshka + normal
reward model -> RewardPoolerHead -> softmax

Actually, the two don’t share much common code.

**Comment by DarkLight1337** [comment]
So you mean we have a completely separate pooler head defined for each model type?

**Comment by DarkLight1337** [comment]
In that case we should call the one for embedding model `EmbeddingPoolerHead`

**Comment by noooop** [comment]
@DarkLight1337 

Considering the reward models, the situation is much more complex than I imagined.

To make the control logic simpler and more intuitive, we need to refactor the entire pooler control logic as follows:

override_pooler_config -> pooler_config -> PoolingParams -> Pooler( PoolingMethod& PoolerActivation)

1. pooler_config & override_pooler_config control during startup. 

2. PoolingParams control per request

PoolingParams needs to merge the default parameters from pooler_config

PoolerActivation needs to support per request control

3. To achieve control per request, we need to load act_fn at any time, and determine at runtime whether to use it.

4. Control Logic

4.1 pooling_type controls Pooler.pooling, I think it can only be modified at startup.

4.2 the control parameters of embed, classify, reward are completely non-overlapping.

4.2.1 for embed models, 

Consider only the dimensions and normalize parameters

4.2.2 for classification (score) models

Consider only the activation parameter

4.2.3 for reward models

4.2.3.1 for PoolingType.ALL 

Consider only the softmax parameter

4.2.3.2 for PoolingType.STEP 
Consider only the softmax/step_tag_id/returned_token_ids parameter

***

┓( ´∀` )┏


Are you really going to change this mess?

**Comment by DarkLight1337** [comment]
Maybe we can have something like `ActivationSelector` which is passed to each pooler to select which activation to use based on the PoolingParams and the original config

**Review by maxdebayser — APPROVED** [review]
LGTM

File: vllm/model_executor/layers/pooler.py [hunk 5]
```diff
@@ -541,7 +512,48 @@ def forward(self, pooled_data: Union[list[torch.Tensor], torch.Tensor],
...
+
+
+class RewardPoolerHead(PoolerHead):
+
+    def __init__(self) -> None:
+        super().__init__(activation=PoolerClassify())
+
+    def forward(self, pooled_data: Union[list[torch.Tensor], torch.Tensor],
+                pooling_metadata: PoolingMetadata):
+        # Using float32 in PoolerHead
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py]: Doesn't the activation function already cast to the correct dtype?

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py]: I think there is no need for pooler head to handle dtype casting.

File: vllm/model_executor/layers/pooler.py [hunk 6]
```diff
@@ -559,8 +571,12 @@ def from_config(
         pooler_config: ResolvedPoolingConfig,
     ) -> "SimplePooler":
         pooling = PoolingMethod.from_pooling_type(pooler_config.pooling_type)
-        head = PoolerHead.from_config(pooler_config)
-
+        if pooler_config.task == "embed":
+            head = EmbeddingPoolerHead()
+        elif pooler_config.task == "encode":
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:553]: I think this breaks Qwen2-RM (not PRM) which uses last pooling with identify activation

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:553]: I'm not sure if adding Qwen2ForRewardModelConfig can solve the issue.

```
class Qwen2ForRewardModelConfig(VerifyAndUpdateConfig):

    @staticmethod
    def verify_and_update_config(vllm_config: "VllmConfig") -> None:
        pooler_config = vllm_config.model_config.pooler_config

        if pooler_config.softmax is None:
            pooler_config.softmax = False
```

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:553]: I don't think that'll work, since the reward head has a fixed activation

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:553]: However, this would set default PoolingParams.softmax = False instead of PoolingParams.softmax = None by set default model_config.pooler_config.softmax = False nstead of model_config.pooler_config.softmax = None

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:553]: config priority, from low to high
1. vllm default in vllm/pooling_params.py _set_default_parameters
2. get_pooling_config (from sentencetransformer config) in vllm/config.py _init_pooler_config
3. verify_and_update_config in vllm/model_executor/models/config.py
4. override_pooler_config in vllm/config.py _init_pooler_config
5. PoolingParams.__init__ in vllm/pooling_params.py

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:553]: I think you need to add some documentatoin on how this works now

File: vllm/model_executor/layers/pooler.py [hunk 7]
```diff
@@ -541,7 +498,41 @@ def forward(self, pooled_data: Union[list[torch.Tensor], torch.Tensor],
...
+        else:
+            pooled_data = [
+                self.activation(vecs) if f else vecs
+                for vecs, f in zip(pooled_data, flags)
+            ]
+
+        return pooled_data
+
+
+class RewardPoolerHead(PoolerHead):
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:513]: Also I think we should just call this `ClassificationPoolerHead`

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:513]: Since reward models might not actually use this

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:513]: Now Pooler is always loaded the activation, and we can control whether it is used per request by PoolingParams

- EmbeddingPoolerHead + SimplePooler for embed models,
- ClassifierPooler for classification (score) models
- RewardPoolerHead + SimplePooler for PoolingType.ALL reward models
- RewardPoolerHead + StepPooler for PoolingType.STEP reward models

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:513]: Oh I see, the activation is skipped because you set `softmax`...

File: vllm/model_executor/layers/pooler.py (hunk 6 above)
Comment (noooop) [reply vllm/model_executor/layers/pooler.py:553]: In fact, this is easy to understand. 
Now the Pooler is only controlled by PoolingParams, so the initialization parameters of PoolingParams (User input per request) have the highest priority. 

If the parameters are None, they will fail back to override_pooler_config, verify_and_update_config, get_pooling_config, and vllm default in sequence.

Comment (noooop) [reply vllm/model_executor/layers/pooler.py:553]: The document is too difficult for me. 

I'll write in a separate pr one once this feature stabilizes.

**Review by DarkLight1337 — APPROVED** [review]
Let's see if the tests pass

File: tests/entrypoints/llm/test_score.py (hunk 1 above)
Comment (noooop) [reply tests/entrypoints/llm/test_score.py:18]: @DarkLight1337 

I don't know how to use this with only v0, 

removing def v1(run_with_both_engines): or the one below @pytest.mark.skip_v1 doesn't work either.

File: tests/entrypoints/llm/test_score.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+              gpu_memory_utilization=0.75,
+              enforce_eager=True,
+              seed=0)
+
+    with llm.deprecate_legacy_api():
+        yield weakref.proxy(llm)
+
+        del llm
+
+    cleanup_dist_env_and_memory()
```
Comment (noooop) [inline tests/entrypoints/llm/test_score.py]: @DarkLight1337

I don't know how to use this with only v0,

removing def v1(run_with_both_engines): or the one below @pytest.mark.skip_v1 doesn't work either.

Comment (DarkLight1337) [reply tests/entrypoints/llm/test_score.py]: Maybe you can explicitly check `envs.VLLM_USE_V1` and skip the test using `pytest.skip`

**Comment by noooop** [comment]
@DarkLight1337 

Compare current with merging main into this PR

https://buildkite.com/vllm/ci/builds/25929/steps/canvas?sid=0198745d-f9c5-4e6a-87a6-7b674c571a02


Can fix the following error
- buildkite/ci/pr/entrypoints-test-api-server
- buildkite/ci/pr/openai-api-correctness
- buildkite/ci/pr/samplers-test


These error can all be fixed by #22088

Image download issue
- buildkite/ci/pr/multi-modal-models-test-standard 

But it will cause new errors
- [Language Models Test (Extended Pooling)](https://buildkite.com/vllm/ci/builds/25929/steps/canvas?sid=0198745d-f9c5-4e6a-87a6-7b674c571a02)

**Comment by DarkLight1337** [comment]
Can you fix that error then?

**Comment by noooop** [comment]
I think this is a CI bug

I used exactly the same code

**Comment by DarkLight1337** [comment]
Does the error happen on main branch without this PR?

**Comment by noooop** [comment]
No error  yesterday https://buildkite.com/vllm/ci/builds/25855/steps/canvas

**Comment by DarkLight1337** [comment]
Then we need to avoid introducing the error before merging this PR. Can you merge from main branch and run the CI again?

**Comment by noooop** [comment]
@DarkLight1337 

Language Models Test (Extended Pooling) pass. Luck
